### ISSUE TITLE
due to NCCL requirement, will not work on Windows

### ISSUE THREAD

**Issue by ThomasDotCodes** [genforce/idinvert#25]
Because of no NCCL support for windows, the dependencies of this project prevent it from running on Windows:
https://github.com/tensorflow/tensorflow/issues/21470

This _appears_ to be the only place that nccl is used, and from what I understand (could be wrong) is that NCCL is only needed for multiple-gpu operations anyways.
https://github.com/genforce/idinvert/blob/17cbd86ddfbbab3b231b961d4c3a8d0640625e71/dnnlib/tflib/optimizer.py#L135

Any ideas on how to remove the NCCL requirement to allow this project to run in Windows?

**Comment by zhujiapeng** [comment]
The NCCL is the basis for running this code on multi-GPUs, you could run the code on a single GPU by removing it, but I am not sure whether the performance on a single GPU could reach the performance on multi-GPUs.

**Comment by ThomasDotCodes** [comment]
Thank you, and understood.  In a practical sense, I assume I could just remove this then, and not replace it with anything? (if using a single GPU):

https://github.com/genforce/idinvert/blob/17cbd86ddfbbab3b231b961d4c3a8d0640625e71/dnnlib/tflib/optimizer.py#L134-L135

**Comment by zhujiapeng** [comment]
Actually, if you use a single GPU, the code can not reach here. By the way, if you use a single GPU to training the encoder, you need to replace the `sync_batch_norm` in `networks_encoder.py` with conventional BatchNorm provided by TensorFlow.
![code](https://user-images.githubusercontent.com/10594850/96950531-17a73100-151d-11eb-8a8e-6de5b08c95c0.png)

**Comment by ThomasDotCodes** [comment]
Thanks for the advice so far, looks like the import in `optimizer.py` wasn't the issue

In fact, I've removed all references to nccl as found here:
https://github.com/genforce/idinvert/search?q=nccl

yet still continue to get nccl related errors
![image](https://user-images.githubusercontent.com/281948/97083412-c612ab00-15dd-11eb-8e1f-7207f54c4e5d.png)

It seems to be from here, but I'm not sure how to trace the issue further
https://github.com/genforce/idinvert/blob/master/dnnlib/tflib/network.py#L291

**Comment by zhujiapeng** [comment]
The `.pkl` files contain the network's code I used in training and when it was loaded, the code in it will be directly used.

**Comment by ThomasDotCodes** [comment]
Thank you for the explanation, that makes sense.  Closing this as an issue.
